### PR TITLE
Parallelise DASSL and IDA Jacobian evaluation

### PR DESCRIPTION
**Pull request by AnHeuermann**
(Includes PR #302 for testing purpose)

# Parallelise DASSL and IDA Jacobian evaluation
Add support for parallel Jacobian evaluation in DASSL and IDA for symbolical jacobian of right hand side of DAE.

## Purpose
Speed up simulation time with faster jacobian evaluations. Could make a difference for models with big jacobians where jacobian evaluation makes up a significant time ot overall simulation time.

## Usage
- Parallel Jacobian evaluation will be encapsulated by USE_PARJAC defines.
- To compile omc with parallel jacobians pass `--enable-parjac` to `OMCompiler/configure`. This will check for OpenMP support and add corresponding Flags to simulation makefile and c runtime makefile.
- To simulate a model with parallel jacobian evaluations use
  a) Pass `-jacobianThreads=<numberOfThreads>` as simulationflag to use desired ammount of threads for parallel jacobian evaluation.
  b) environment variable `OMP_NUM_THREADS=N`
  This order also gives the order of the precedence, i.e. if both are specified value of -jacobianThreads is taken over `OMP_NUM_THREADS`. If nothing is specified by the user `omp_get_max_thread()` is used to set the number of threads to use for parallel jacobian evaluation.

## Changes
- Added function to allocate and deallocate memory for parallel jacobian structure used by IDA and DASSL in `jacobianSymbolical.c`
- Do the jacobian evaluation in function `jacA_sym` in parallel.
- Linear systems can now be computed in parallel by use of thread safe data structure.
  - New structure `struct LINEAR_SYSTEM_THREAD_DATA` to strore data for each thread.
  - Changed way to allocate linear system data with new `parDynamicData`. This is an array of length of number of threads to be used. When `enable-parjac=no` it is length 1.
- Added some helper functions to get number of current thread, maximum number of threads and to work with and without OpenMP.

## ToDo in next PR
There are still things to do but I'd like to do these in an other PR.
- Allocate memory for linear systems only thread-many often for systems that are part of Jacobian A of DASSL/IDA. Currently this is done for all systems.
- Include testcases / CI.
- Include Windows build.
- Handle stream prints from parallel regions correctly.

### CODE REVIEW

**Comment by AnHeuermann** [comment]
@sjoelund Do you know a good way to add a test without changing the default configure options?

One needs to reconfigure with `--enable-parjac` and rebuild parts of c runtime to simulate a model with `-jacobianThreads=<numberOfThreads>`

**Comment by sjoelund** [comment]
A new Jenkinsfile, like it is done for .CI/Jenkinsfile.cpp or .CI/Jenkinsfile.more-compilers. Or a new optional target like it is done for MSVC and OSX.

Better would be to not require a different configure option, but to compile additional solvers only if OMP is available. Maybe something like:
dassl_serial.c
```
#define DASSL_FUNCTION_PREFIX(X) omc_serial ## X
#include "dassl.c"
```

```
#define DASSL_FUNCTION_PREFIX(X) omc_parallel ## X
#include <omp.h>
#include "dassl.c"
```

And then have potentially both in the solvers you can choose from (dassl.c would contain the common code and ifdef some parts depending on if omp is used).

File: OMCompiler/SimulationRuntime/c/simulation/solver/linearSystem.c [hunk 1]
```diff
@@ -396,8 +396,10 @@ int freeLinearSystems(DATA *data, threadData_t *threadData)
 int solve_linear_system(DATA *data, threadData_t *threadData, int sysNumber, double* aux_x)
 {
   TRACE_PUSH
-  int success;
   int retVal;
+#pragma omp critical
+{
+  int success;
```
Comment (AnHeuermann) [inline OMCompiler/SimulationRuntime/c/simulation/solver/linearSystem.c]: This solves the problems regarding garbage collector and systems with many linear systems. It's not prity but in this PR I want only to parallelize the Jacobian evaluation inside DASSL and IDA.

**Comment by AnHeuermann** [comment]
I would remove some temporary changes and leave the compilation of omc with the configure of `--enable-parjac`. Some models inside the ScalableTestSuite can be simulated notable faster, but I want to see what happens in general bevor making it a default option.

**Comment by AnHeuermann** [comment]
As you can see in [this test](https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-361/30/pipeline) where everything is build with parallelization and the maximum nuber of threads possible used as default only tests for FMI/OMSimulator and parmodelica are failing.
So I would suggest merging this PR after the Beta release. Please give some feedback on the implementation.

Deciding if we want to use the parallelization as default should be done in a different PR.

**Comment by AnHeuermann** [comment]
@adrpo How could I change the build process on Windows to configure with `--enable-parjac` like on Unix? Like in [this commit](https://github.com/OpenModelica/OpenModelica/pull/361/commits/385ded6c0c93ba6e56158f38aa3e4166a4f31273) where it is done for Unix.
I need to update `RUNTIMECFLAGS="$RUNTIMECFLAGS $OMPCFLAGS -DUSE_PARJAC"` but haven't found the right location in the MinGW Makefiles.

**Pull request merged**